cvxpy/tests/solver_test_helpers.py

```diff
@@ -110,27 +110,116 @@ def check_complementarity(self, places) -> None:
-                                  cp.constraints.Zero)):
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.Zero,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
                 comp = cp.scalar_product(con.args, con.dual_value).value
-            elif isinstance(con, cp.constraints.PowCone3D):
-                comp = cp.scalar_product(con.args[:3], con.dual_value).value
-            elif isinstance(con, (cp.constraints.Inequality,
-                                  cp.constraints.Equality)):
-                comp = cp.scalar_product(con.expr, con.dual_value).value
             elif isinstance(con, cp.constraints.PowConeND):
                 msg = '\nPowConeND dual variables not implemented;' \
                        + '\nSkipping complementarity check.'
                 warnings.warn(msg)
             else:
                 raise ValueError('Unknown constraint type %s.' % type(con))
             self.tester.assertAlmostEqual(comp, 0, places)
+            
+    def check_stationary_lagrangian(self, places) -> None:
+        L = self.prob.objective.expr
+        psd_flag = False
+        for con in self.constraints:
+            if isinstance(con, (cp.constraints.Inequality,
+                                cp.constraints.Equality)):
+                dual_var_value = con.dual_value
+                prim_var_expr = con.expr
+                L = L + cp.scalar_product(dual_var_value, prim_var_expr)
+            elif isinstance(con, (cp.constraints.ExpCone,
+                                  cp.constraints.SOC,
+                                  cp.constraints.Zero,
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
+                L = L - cp.scalar_product(con.args, con.dual_value)
+            else:
+                raise NotImplementedError()
+        g = L.grad
+
+        # compute norm
+        bad_fro_norms = []
+        for (k, v) in g.items():
+            # (k, v) = (cvxpy Variable, SciPy sparse matrix)
+            norm = np.linalg.norm(v.data) / np.sqrt(k.size)
+            if norm > 10**(-places):
+                bad_fro_norms.append((norm, k))
+
+        # Attempt at "correcting" 'bad' norms:
+        if len(bad_fro_norms):
+            """The reason 'bad' norms could arise (outside of an error by CVXPY) would be if a
+            constraint was introduced in a variable via a flag during it's declaration and not
+            explicitly in the list of constraints passed to the problem --- thus making the above
+            constructed lagrangian incorrect. In these cases instead of checking if the dLdX = 0
+            we check if dLdX \in K^{*} (i.e. the dual cone of the implicit constraint)"""
+            for opt_var in self.prob.variables()
```

Comment:
Question: Should this pass through the problem variables come before or after the computation of `bad_fro_norms` computation?
If we can figure out a way to compute a `norm` $\to 0$ for every type of 'flag' constraint, then it should definitely be the former

Reply:
For `diag` this could just be `np.linalg.norm(np.diag(opt_var.toarray()))`.

For `nonneg` it could be `np.linalg.norm(opt_var.toarray() - np.abs(opt_var.to_array()))`
Similarly, for `nonneg` it could be `np.linalg.norm(opt_var.toarray() + np.abs(opt_var.to_array()))`

But I am not sure what it could be a norm definition which goes to zero in case a matrix is `PSD`.

Reply:
On @phschiele's suggestion, a good candidate for a quantity in this case would just be the difference between the projection of the matrix variable onto the PSD cone and the matrix itself. 

This is obviously, already done within `residual`! We could also use the same quantity in the case of `nonneg` and `nonpos` since they are also self-dual. My suggestion for `diag` could still be used though.